tests/end-to-end-tests.spec.ts

```diff
@@ -0,0 +1,268 @@
+        alice.did,
+        protocolsConfigureForAlice.message
+      );
+      expect(protocolsConfigureForAliceReply.status.code).to.equal(202);
+
+      // Bob configures chat protocol with encryption
+      const protocolDefinitionForBob
+        = await Protocols.deriveAndInjectPublicEncryptionKeys(protocolDefinition, bob.keyId, bob.keyPair.privateJwk);
+      const protocolsConfigureForBob = await TestDataGenerator.generateProtocolsConfigure({
+        author             : bob,
+        protocolDefinition : protocolDefinitionForBob
+      });
+
+      const protocolsConfigureReply = await dwn.processMessage(bob.did, protocolsConfigureForBob.message);
+      expect(protocolsConfigureReply.status.code).to.equal(202);
+
+      // 1. Alice starts a chat thread writing to her own DWN
+      const threadBytes = Encoder.objectToBytes({ title: 'Top Secret' });
+      const threadRecord = await TestDataGenerator.generateProtocolEncryptedRecordsWrite({
+        plaintextBytes                                   : threadBytes,
+        author                                           : alice,
+        protocolDefinition                               : protocolDefinition,
+        protocolPath                                     : 'thread',
+        encryptSymmetricKeyWithProtocolPathDerivedKey    : false,
+        encryptSymmetricKeyWithProtocolContextDerivedKey : true
+      });
+      const threadRecordReply1 = await dwn.processMessage(alice.did, threadRecord.message, threadRecord.dataStream);
+      expect(threadRecordReply1.status.code).to.equal(202);
+
+      // 2. Alice adds Bob as a participant giving him the [context-derived private key] encrypted using [Bob's thread-level public key]
+
+      // the context-derived key to be used for encrypting symmetric keys
+      const aliceRootKey = {
+        rootKeyId         : alice.keyId,
+        derivationScheme  : KeyDerivationScheme.ProtocolContext,
+        derivedPrivateKey : alice.keyPair.privateJwk
+      };
+      const contextDerivationPath = Records.constructKeyDerivationPathUsingProtocolContextScheme(threadRecord.message.contextId);
+      const contextDerivedPrivateKey: DerivedPrivateJwk = await HdKey.derivePrivateKey(aliceRootKey, contextDerivationPath);
+      const contextDerivedPublicKey = threadRecord.encryptionInput.keyEncryptionInputs[0].publicKey;
+
+      // Alice queries Bob's DWN for Bob's chat protocol definition containing public key declarations
+      const protocolsQuery = await ProtocolsQuery.create({
+        filter: { protocol: threadRoleProtocolDefinition.protocol }
+      });
+      const protocolsQueryReply = await dwn.processMessage(bob.did, protocolsQuery.message);
+      const protocolConfigureMessageOfBobFetched = protocolsQueryReply.entries![0] as ProtocolsConfigureMessage;
+
+      // Alice verifies that the chat protocol definition is authored by Bob
+      await authenticate(protocolConfigureMessageOfBobFetched.authorization, didResolver);
```

Comment:
Might be useful to export `authenticate` in `index.ts` since it's necessary for this e2e example.

Reply:
Done!